run-build-functions.sh

```diff
@@ -102,6 +107,33 @@ run_yarn() {
   fi
 
 
+  if [ "$NETLIFY_YARN_WORKSPACES" = "true" ]
+  then
+    echo "NETLIFY_YARN_WORKSPACES feature flag set"
+    # YARN_IGNORE_PATH will ignore the presence of a local yarn executable (i.e. yarn 2) and default
+    # to using the global one (which, for now, is alwasy yarn 1.x). See https://yarnpkg.com/configuration/yarnrc#ignorePath
+    if YARN_IGNORE_PATH=1 yarn workspaces info
+    then
+      # Extract all the packages and respective locations. .data will be a JSON object like
+      # {
+      #   "my-package-1": {
+      #     "location": "packages/blog-1",
+      #     "workspaceDependencies": [],
+      #     "mismatchedWorkspaceDependencies": []
+      #   },
+      #   (...)
+      # }
+      # We need to cache all the node_module dirs, or we'll always be installing them on each run
+      local package_locations=($(YARN_IGNORE_PATH=1 yarn --json workspaces info | jq -r '.data | fromjson | to_entries | .[].value.location | @sh'| tr -d \'))
```

Comment:
Note: I do not believe this works when there are spaces in file paths.
This is because `$()` is not quoted. However quoting it would not work with array creation `( )`.

Reply:
Good call @ehmicky, we should handle file paths with spaces. This is probably a sign that we should do this another way to avoid bash sourceries related to word splitting also. I searched around a bit and from a page of `jq` faq - https://github.com/stedolan/jq/wiki/FAQ - we should probably rely on something like `mapfile -t array <<< $(jq -c '.[]' input.json)`

Reply:
Handled via https://github.com/netlify/build-image/commit/ef73c8409763fd955bc04f8e1c1d1cb58e99a2b1

Running it locally:
```sh
NETLIFY_YARN_WORKSPACES feature flag set
yarn workspaces v1.22.4
{
  "gatsby-starter-blog-1": {
    "location": "packages/blog-1",
    "workspaceDependencies": [],
    "mismatchedWorkspaceDependencies": []
  },
  "gatsby-starter-blog-2": {
    "location": "packages/blog-2",
    "workspaceDependencies": [],
    "mismatchedWorkspaceDependencies": []
  },
  "blog-3": {
    "location": "packages/dir with spaces",
    "workspaceDependencies": [],
    "mismatchedWorkspaceDependencies": []
  }
}
Done in 0.12s.
Started restoring workspace packages/blog-1 node modules
Finished restoring workspace packages/blog-1 node modules
Started restoring workspace packages/blog-2 node modules
Finished restoring workspace packages/blog-2 node modules
Started restoring workspace packages/dir with spaces node modules
Finished restoring workspace packages/dir with spaces node modules
Started restoring workspace root node modules
Finished restoring workspace root node modules

(...)

Caching artifacts
Started saving workspace packages/blog-1 node modules
Finished saving workspace packages/blog-1 node modules
Started saving workspace packages/blog-2 node modules
Finished saving workspace packages/blog-2 node modules
Started saving workspace packages/dir with spaces node modules
Finished saving workspace packages/dir with spaces node modules
Started saving workspace root node modules
Finished saving workspace root node modules

(...)
```

Reply:
:tada: 

Would `|` work instead of `<<< "$(...)"`? 

```bash
YARN_IGNORE_PATH=1 yarn --json workspaces info | jq -r '.data | fromjson | to_entries | .[].value.location' | mapfile -t package_locations
```

Reply:
Piping won't work here unfortunately as it would create subshells to execute each part of it, meaning any changes to the `package_locations` variable will be local to that subshell - https://stackoverflow.com/a/18116407

Reply:
Oh, that's correct! :+1: